### PR TITLE
Release 1.0.0

### PR DESCRIPTION
**Pull request by x-tabdeveloping**
Changed the API completely to make the library more comfortable to use.
Main changes:
### `Trooper`
The brand new `Trooper` interface allows you not to have to specify what model type you wish to use.
Stormtrooper will automatically detect the model type from the specified name.

```python
from stormtrooper import Trooper

# This loads a setfit model
model = Trooper("all-MiniLM-L6-v2")

# This loads an OpenAI model
model = Trooper("gpt-4")

# This loads a Text2Text model
model = Trooper("google/flan-t5-base")
```

### Unified zero and few-shot classification

You no longer have to specify whether a model should be a few or a zero-shot classifier when initialising it.
If you do not pass any training examples, it will be automatically assumed that the model should be zero-shot.

```python
# This is a zero-shot model
model.fit(None, ["dog", "cat"])

# This is a few-shot model
model.fit(["he was a good boy", "just lay down on my laptop"], ["dog", "cat"])
```

### Reimplemented SetFit
The SetFit library caused me many headaches, and has unfixed bugs, and I didn't want to downgrade everything.
I implemented the SetFit algorithm from scratch using training utilities in sentence-transformers.

### New docs in MKDocs
Switched ti MKDocs, because it's simpler to deal with and looks better.

### Test suite
Added tests to ensure that the library will not break overnight.

### Multiple GPU support
Run models on multiple GPU's by passing: `Trooper(model_name, device_map="auto")`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,32 +7,45 @@ Zero/few shot learning components for scikit-learn pipelines with large-language
...
-   - Throrough API reference and loads of examples to get you started.
-3. Battle-hardened
-    - We at the Center For Humanities Computing are making extensive use of this package. This means you can rest assured that the package works under real-world pressure. As such you can expect regular updates and maintance.
-4. Simple
-    - We opted for as bare-bones of an implementation and little coupling as possible. The library works at the lowest level of abstraction possible, and we hope our code will be rather easy for others to understand and contribute to.
+```python
+from stormtrooper import Trooper
+
+# This loads a setfit model
+model = Trooper("all-MiniLM-L6-v2")
```
Comment (KennethEnevoldsen) [inline README.md:20]: I like the general interface, but would probably call it "AutoClassifier" or similar.

How does it get the model here - doesn't it need "sentence-transformer/*"?

File: README.md [hunk 2]
```diff
@@ -7,32 +7,45 @@ Zero/few shot learning components for scikit-learn pipelines with large-language
...
 
+# This loads an OpenAI model
+model = Trooper("gpt-4")
```
Comment (KennethEnevoldsen) [inline README.md:23]: How do I know which models are available?

File: README.md [hunk 3]
```diff
@@ -7,32 +7,45 @@ Zero/few shot learning components for scikit-learn pipelines with large-language
...
+
+```python
+# This is a zero-shot model
+model.fit(None, ["dog", "cat"])
+
+# This is a few-shot model
+model.fit(["he was a good boy", "just lay down on my laptop"], ["dog", "cat"])
+```
+
+## New in 0.5.0
```
Comment (KennethEnevoldsen) [inline README.md]: Since this is new in this PR, this might seen misleading

File: README.md [hunk 4]
```diff
@@ -41,83 +54,49 @@ Text to clasify:
...
+4. OpenAI models e.g. `Trooper("gpt-4")`
+5. NLI models e.g. `Trooper("facebook/bart-large-mnli")`
+
+## Example usage
+
+Find more in our [docs](https://centre-for-humanities-computing.github.io/stormtrooper/).
 
 ```bash
-pip install stormtrooper
+pip install stormtrooper[torch]
```
Comment (KennethEnevoldsen) [inline README.md]: Why not just include torch by default? (or not specify it here)

File: README.md [hunk 5]
```diff
@@ -133,5 +112,14 @@ From version 0.2.2 you can run models on GPU.
...
+classifier = Trooper("all-MiniLM-L6-v2", device="cuda:0")
+```
+
+### Inference on multiple GPUs
+
+You can run a model on multiple devices in order of device priority `GPU -> CPU + Ram -> Disk` and on multiple devices by using the `device_map` argument.
+Note that this only works with text2text and generative models.
+
+```
+model = Trooper("HuggingFaceH4/zephyr-7b-beta", device_map="auto")
```
Comment (KennethEnevoldsen) [inline README.md:105]: This seems a bit odd to me - why not just pass device?

File: README.md [hunk 6]
```diff
@@ -41,83 +54,49 @@ Text to clasify:
...
+# ["atheism/christianity", "astronomy/space"]
+
+# Few-shot classification
+model = Trooper("google/flan-t5-base")
+model.fit(example_texts, class_labels)
+model.predict(new_texts)
+# ["atheism/christianity", "astronomy/space"]
 ```
 
 ### Fuzzy Matching
```
Comment (KennethEnevoldsen) [inline README.md:83]: So it does this generally?

File: mkdocs.yml
Comment (KennethEnevoldsen) [inline mkdocs.yml:1]: Seems like you have a lot of docs in the readme - I would probably keep it in either (minimal example in readme - everything else in docs)

File: stormtrooper/set_fit.py [hunk 7]
```diff
@@ -1,209 +1,96 @@
...
-        Class names learned from the labels.
+        Device to train and run the model on.
+    classification_head: ClassifierMixin, default None
+        Classifier to use as the last step.
+        Defaults to Logistic Regression when not specified.
+    n_epochs: int, default 10
+        Number of trainig epochs.
+    batch_size: int, default 8
+        Batch size to use during training.
+    sample_size: int, default 8
```
Comment (KennethEnevoldsen) [inline stormtrooper/set_fit.py:27]:
```suggestion
    sample_size
```
Generally I would not specify defaults or types in docs (then it is specified twice).

File: tests/test_all.py
Comment (KennethEnevoldsen) [inline tests/test_all.py:1]: rename test_trooper.py

File: stormtrooper/trooper.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+        The device the model should run on (only valid for locally run models).
+    device_map: str, default None
+        Device map argument for very large models.
+    prompt: str, default None
+        Prompt to use for promptable models.
+    system_prompt: str, default None
+        System prompt to use for chat models.
+    fuzzy_match: bool, default True
+        Indicates whether responses should be fuzzy matched to the closest class name,
+        when using a generative model.
```
Comment (KennethEnevoldsen) [inline stormtrooper/trooper.py:107]: So e.g. if it misspells a case (pstive) - or ?

**Review by KennethEnevoldsen — APPROVED** [review]
Next time I would probably split the pr into multiple.

CI:
- It might be a good idea to make a makefile to unify CI across both local and CI

I would like to know more about how it selects models. - Mixing the APIs with HF models is also a bit confusing. I would split it up.

You are also hiding a lot of complexity from the user, it might be worth letting people know what is going on below the hood.

File: README.md (hunk 5 above)
Comment (x-tabdeveloping) [reply README.md:105]: This is not me, but HuggingFace.

File: README.md (hunk 6 above)
Comment (x-tabdeveloping) [reply README.md:83]: ? I'm not sure I understand

File: stormtrooper/set_fit.py (hunk 7 above)
Comment (x-tabdeveloping) [reply stormtrooper/set_fit.py:27]: I'm using Numpy style docstrings with a plugin, and all of the code was written this way because the Sphinx plugin needed it, it's unnecessary, but I think it's fine for now.

File: stormtrooper/trooper.py (hunk 8 above)
Comment (x-tabdeveloping) [reply stormtrooper/trooper.py:107]: Yeah, so if the match is not exact from the language model, then this matches it to the closest option.

File: README.md (hunk 5 above)
Comment (KennethEnevoldsen) [reply README.md:105]: Then I would just say that you can pass kwargs onto the HF model. This sounds like you are specifying the interface.
